pkg/controllers/core/deployer_controller.go

```diff
@@ -557,6 +558,38 @@ func (r *DeployerReconciler) constructServiceForDeployer(deployer *corev1alpha1.
 	// If a service name was provided, use it, otherwise generate one based on the deployer name.
 	if deployer.Spec.ServiceName != "" {
 		service.ObjectMeta.Name = deployer.Spec.ServiceName
+
+		// Detect service name collision.
+		collidingService, err := r.getService(deployer.Namespace, deployer.Spec.ServiceName)
+		if err != nil {
+			return nil, err
+		}
+		if collidingService != nil {
+			// Diagnose service name collisions. If the colliding service was created by a deployer, include that in the
+			// diagnostics.
+			collisionKind := collidingService.Kind
+			collisionName := collidingService.Name
+
+			for _, ownerRef := range collidingService.OwnerReferences {
+				if ownerRef.Kind == deployer.Kind {
+					collisionKind = deployer.Kind
+					collisionName = ownerRef.Name
+					break
+				}
+			}
+			err := fmt.Errorf("service name provided on deployer collided with existing service %v", collidingService)
+			r.Log.Error(err, "collision", "deployer", deployer, "collidingService", collidingService,
+				"collisionKind", collisionKind, "collisionName", collisionName)
+			deployer.Status.ServiceNameCollision(collisionKind, collisionName)
+
+			// Track the colliding service for collision removal.
+			r.Tracker.Track(
```

Comment:
tracking the collided resource is an anti-pattern in this case. If that resource is part of a production system, imagine an upgrade needs to delete and then recreate the same resource (some fields are immutable). If this controller is watching for the resource to be deleted, we now have a race for who can create the resource first.

Reply:
Good point about non-determinism. But if we don't track the collided resource, how will we know when to re-reconcile if the collided resource if modified or deleted to remove the collision?